tasks/devkit.rake

```diff
@@ -1,118 +1,173 @@
-  env = TDK::Environment.new(args.env)
-  env.local_backend? || (raise 'Testing is only allowed for local environments')
-
-  task('apply').invoke(env.name)
-
-  destroy_if_fails(env) do
-    invoke_if_defined('custom_test', args.env)
-  end
-end
-
-desc 'Creates the infrastructure and run the tests'
-task :preflight, [:teardown] do |_, args|
-  args.with_defaults(teardown: 'true')
-  env = TDK::Environment.new(TDK::Environment.temp_name)
-  task('test').invoke(env.name)
-  task('clean').invoke(env.name) if args.teardown == 'true'
-end
-
-desc 'Destroys the infrastructure'
-task :destroy, [:env] => :prepare do |_, args|
-  invoke_if_defined('pre_destroy', args.env)
-
-  env = TDK::Environment.new(args.env)
-  cmd = 'terragrunt destroy'
-  cmd += ' -force' if env.local_backend?
-  TDK::Command.run(cmd, directory: env.working_dir, close_stdin: false)
-
-  invoke_if_defined('post_destroy', args.env)
-end
-
-desc 'Cleans an environment (infrastructure is destroyed too)'
-task :clean, [:env] => :destroy do |_, args|
-  env = TDK::Environment.new(args.env)
-  puts "Deleting environment #{env.name}"
-  FileUtils.rm_rf(env.working_dir, secure: true)
-end
+require 'fileutils'
+require 'TerraformDevKit'
+
+TDK = TerraformDevKit
+
+raise 'ROOT_PATH is not defined' if defined?(ROOT_PATH).nil?
+BIN_PATH = File.join(ROOT_PATH, 'bin')
+
+# Ensure terraform is in the PATH
+ENV['PATH'] = TDK::OS.join_env_path(
+  TDK::OS.convert_to_local_path(BIN_PATH),
+  ENV['PATH']
+)
+
+PLAN_FILE = 'plan.tfplan'.freeze
+
+def destroy_if_fails(env)
+  yield
+rescue StandardError => e
+  puts "ERROR: #{e.message}"
+  puts e.backtrace.join("\n")
+  task('destroy').invoke(env.name) if env.local_backend?
+  raise
+end
+
+def invoke_if_defined(task_name, env)
+  task(task_name).invoke(env) if Rake::Task.task_defined?(task_name)
+end
+
+def lock_table
+  aws_config = TDK::AwsConfig.new(TDK::Configuration.get('aws'))
+  dynamo_db = TDK::DynamoDB.new(
+    aws_config.credentials,
+    aws_config.region
+  )
+  s3 = TDK::S3.new(
+    aws_config.credentials,
+    aws_config.region
+  )
+  TDK::TerraformLockTable.new(dynamo_db, s3)
+end
+
+desc 'Prepares the environment to create the infrastructure'
+task :prepare, [:env] do |_, args|
+  puts "== Configuring environment #{args.env}"
+  env = TDK::Environment.new(args.env)
+
+  config_file = "config/config-#{env.config}.yml"
+  puts "== Loading configuration from #{config_file}"
+  TDK::Configuration.init(config_file)
+
+  TDK::TerraformInstaller.install_local(
+    TDK::Configuration.get('terraform-version'),
+    directory: BIN_PATH
+  )
+
+  project_config = TDK::TerraformProjectConfig.new(
+    TDK::Configuration.get('project-name')
+  )
+  TDK::TerraformConfigManager.setup(env, project_config)
+
+  unless env.local_backend?
+    puts '== Initializing remote state'
+    lock_table.create_lock_table_if_not_exists(env, project_config)
+  end
+
+  invoke_if_defined('custom_prepare', args.env)
+
+  TDK::Command.run(
```

Comment:
Do we need to run `init` and `get`? I believe we did it in this way because with terragrunt we couldn't really control when `init` was called. But according to terraform documentation (see next) it might not be needed.

> Usage: terraform init [options] [DIR]
> 
>   Initialize a new or existing Terraform working directory by creating
>   initial files, loading any remote state, downloading modules, etc.
> 
>   This is the first command that should be run for any new or existing
>   Terraform configuration per machine. This sets up all the local data
>   necessary to run Terraform that is typically not committed to version
>   control.
> 
>   This command is always safe to run multiple times. Though subsequent runs
>   may give errors, this command will never delete your configuration or
>   state. Even so, if you have important information, please back it up prior
>   to running this command, just in case.

`init` is certainly required (at least the first time) but `get` might not be necessary.

Reply:
I think it might be fine to just run `init` I'll give it a go.